Catalog update

geomacy opened the pull request:
For AMP-982, work in progress.

Updates BrooklynObject and BrooklynObjectSpec to support nested catalogItemIds.
### TODO:
- doesn't include changes to persistence to allow the nested item ids to be persisted and retrieved during rebind
- Need to review what changes may be needed to EntityTransientCopyInternal.
- Looks like catalogItemIdIfNotNull can be removed.
### Things to review

Everything, really; but some things I made a note of particularly
- review places where setCatalogItemIdIfNotNull has been replaced with nestCatalogItemIds and see if they make sense from a sanity viewpoint
- review use of setCatalogItemIds()

geomacy committed:
Osgi loader change

geomacy committed:
Add nested catalog item ids to AbstractBrooklynObjectSpec.

geomacy committed:
Add catalog item super id support.

geomacy committed:
Fixes to get CatalogYamlEntityTest.testCatalogItemIdInReferencedItems working.

Haven't re-run all unit tests yet.

geomacy commented:
still no good sorry

geomacy commented:
ah got it the commits above are in reverse chronological order; my changes are the last four, as you see.

neykov commented on core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogUtils.java:
```diff
@@ -170,6 +173,13 @@ public static void installLibraries(ManagementContext managementContext, @Nullab
+                BrooklynClassLoadingContext osgiLoader = iterator.next();
+                return getCatalogItemIdFromLoader(osgiLoader);
+            }
+            else return null;
```
Need to revisit this logic in light of latest updates.

neykov commented on core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogItemDtoAbstract.java:
```diff
@@ -45,6 +45,8 @@
 import com.google.common.collect.Sets;
 
+// TODO add support for nested catalog items, implement nestCatalogItemId in terms of symbolicName/Version
+// TODO also getCatalogItemSuperIds.
```
I think we should either override with NOPs, or just keep the default behaviour from `AbstractBrooklynObject`. The catalog items don't know what's the meaning of the plan in them.

neykov reviewed (approved):
Looks good so far.

geomacy committed:
Test with a 'catalog sandwich'.

The purpose of this test is to illustrate that the full stack of nested
catalog items is not searched for classloaders.

geomacy committed:
Do resource lookup from nested catalog ids.

Fixes testDeepCatalogItemCanLoadResources.

geomacy commented:
Retiring this PR in favour of https://github.com/apache/brooklyn-server/pull/338
